<a href="https://colab.research.google.com/github/Rossel/DataQuest_Courses/blob/master/036__List_Comprehensions_and_Lambda_Functions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# COURSE 5/6: DATA CLEANING IN PYTHON: ADVANCED

# MISSION 3: List Comprehensions and Lambda Functions

Learn techniques to turbo-charge working with data in Python

## 1. The JSON Format

So far, we've learned how to use regular expressions to make cleaning and analyzing text data easier.

In this mission, we'll learn some tips and syntax shortcuts we can use on top of everything we've learned, including:

- Creating list comprehensions to replace loops with a single line of code.
- Creating single use functions called lambda functions.

The data set we'll use in this mission is in a format called [JavaScript Object Notation](https://www.json.org/) (JSON). As the name indicates, JSON originated from the JavaScript language, but has now become a language-independent format.

From a Python perspective, JSON can be thought as a collection of Python objects nested inside each other.

![img](https://s3.amazonaws.com/dq-content/355/json.svg)

The JSON above is a list, where each element in the list is a dictionary. Each of the dictionaries have the same keys, and one of the values of each dictionary is itself a list.

The Python `json` [module](https://docs.python.org/3.7/library/json.html#module-json) contains a number of functions to make working with JSON objects easier. We can use the `json.loads()` [method](https://docs.python.org/3.7/library/json.html#json.loads) to convert JSON data contained in a string to the equivalent set of Python objects:

In [ ]:
json_string = """
[
  {
    "name": "Sabine",
    "age": 36,
    "favorite_foods": ["Pumpkin", "Oatmeal"]
  },
  {
    "name": "Zoe",
    "age": 40,
    "favorite_foods": ["Chicken", "Pizza", "Chocolate"]
  },
  {
    "name": "Heidi",
    "age": 40,
    "favorite_foods": ["Caesar Salad"]
  }
]
"""

import json
json_obj = json.loads(json_string)
print(type(json_obj))

<class 'list'>


We can see that `json_string` has turned into a list. Let's take a look at the values in the list:



In [ ]:
print(json_obj)

[{'name': 'Sabine', 'age': 36, 'favorite_foods': ['Pumpkin', 'Oatmeal']}, {'name': 'Zoe', 'age': 40, 'favorite_foods': ['Chicken', 'Pizza', 'Chocolate']}, {'name': 'Heidi', 'age': 40, 'favorite_foods': ['Caesar Salad']}]


We can observe a few things:

- The formatting from our original string is gone. This is because printing Python lists and dictionaries has a simple formatting structure.
- The order of the keys in the dictionary have changed. This is because (prior to version 3.6) Python dictionaries don't have fixed order.

Let's practice using `json.loads()` to convert JSON data from a string to Python objects!



**Instructions:**

We have created a JSON string, `world_cup_str`, which contains data about games from the 2018 Football World Cup.

1. Import the `json` module.
2. Use `json.loads()` to convert `world_cup_str` to a Python object. Assign the result to world_cup_obj.

In [ ]:
# Code to read csv file into Colaboratory:
!pip install -U -q PyDrive
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
from google.colab import auth
from oauth2client.client import GoogleCredentials
# Authenticate and create the PyDrive client.
auth.authenticate_user()
gauth = GoogleAuth()
gauth.credentials = GoogleCredentials.get_application_default()
drive = GoogleDrive(gauth)

In [ ]:
# Once you have completed verification, go to the CSV file in Google Drive, right-click on it and select “Get shareable link”, and cut out the unique id in the link.
# https://drive.google.com/file/d/1J3gYozddULmvWfprsTXK9Ga3V74UpaJe/view?usp=sharing
id = "1J3gYozddULmvWfprsTXK9Ga3V74UpaJe"

In [ ]:
# Download the dataset
downloaded = drive.CreateFile({'id':id}) 
downloaded.GetContentFile('hn_2014.json')

## 2. Reading a JSON file

One of the places where the JSON format is commonly used is in the results returned by an [Application programming interface](https://en.wikipedia.org/wiki/Application_programming_interface) (**API**). APIs are interfaces that can be used to send and transmit data between different computer systems. We'll learn about how to work with APIs in a later course.

The data set from this mission — `hn_2014.json` — was downloaded from the Hacker News API. It's a different set of data from the CSV we've been using in the previous two missions, and it contains data about stories from Hacker News in 2014.

To read a file from JSON format, we use the `json.load()` [function](https://docs.python.org/3.7/library/json.html#json.load). Note that the function is `json.load()` without an "s" at the end. The `json.loads()` function is used for loading JSON data from a string ("loads" is short for "load string"), whereas the `json.load()` function is used to load from a file object. Let's look at how we would read that in our data:



In [ ]:
import json
file = open("hn_2014.json")
hn = json.load(file)

print(type(hn))

<class 'list'>


Our `hn` variable is a list. Let's find out how many objects are in the list, and the type of the first object (which will almost always be the type of every object in the list in JSON data):

In [ ]:
print(len(hn))
print(type(hn[0]))

35806
<class 'dict'>


Our data set contains 35,806 dictionary objects, each representing a Hacker News story. In order to understand the format of our data set, we'll print the keys of the first dictionary:

In [ ]:
print(hn[0].keys())

dict_keys(['author', 'numComments', 'points', 'url', 'storyText', 'createdAt', 'tags', 'createdAtI', 'title', 'objectId'])


If we recall the data set we used in the previous two missions, we can see some similarities. There are keys representing the title, URL, points, number of comments, and date, as well as some others that are less familiar to us. Here is a summary of the keys and the data that they contain:

- `author`: The username of the person who submitted the story.
- `createdAt`: The date and time at which the story was created.
- `createdAtI`: An integer value representing the date and time at which the story was created.
- `numComments`: The number of comments that were made on the story.
- `objectId`: The unique identifier from Hacker News for the story.
- `points`: The number of points the story acquired, calculated as the total number of upvotes minus the total number of downvotes.
- `storyText`: The text of the story (if the story contains text).
- `tags`: A list of tags associated with the story.
- `title`: The title of the story.
- `url`: The URL that the story links to (if the story links to a URL).

Let's start by reading our Hacker News JSON file:



**Instructions:**

The json module remains imported from the previous screen.

1. Use the `open()` [function](https://docs.python.org/3.7/library/functions.html#open) to open the `hn_2014.json` file as a file object.
2. Use the `json.load()` function to parse the file object and assign the result to `hn`.

## 3. Deleting Dictionary Keys

Let's look at the first dictionary in full. To make it easier to understand, we're going to create a function which will print a JSON object with formatting to make it easier to read.

The function will use the `json.dumps()` [function](https://docs.python.org/3.7/library/json.html#json.dumps) ("dump string") which does the opposite of the `json.loads()` function — it takes a JSON object and returns a string version of it. The `json.dumps()` function accepts arguments that can specify formatting for the string, which we'll use to make things easier to read:



In [ ]:
def jprint(obj):
    # create a formatted string of the Python JSON object
    text = json.dumps(obj, sort_keys=True, indent=4)
    print(text)

first_story = hn[0]
jprint(first_story)

{
    "author": "dragongraphics",
    "createdAt": "2014-05-29T08:07:50Z",
    "createdAtI": 1401350870,
    "numComments": 0,
    "objectId": "7815238",
    "points": 2,
    "storyText": "",
    "tags": [
        "story",
        "author_dragongraphics",
        "story_7815238"
    ],
    "title": "Are we getting too Sassy? Weighing up micro-optimisation vs. maintainability",
    "url": "http://ashleynolan.co.uk/blog/are-we-getting-too-sassy"
}


You may notice that the `createdAt` and `createdAtI` keys both have the date and time data in two different formats. Because the format of `createdAt` is much easier to understand, let's do some data cleaning by deleting the `createdAtI` key from every dictionary.

To delete a key from a dictionary, we can use the `del` [statement](https://docs.python.org/3.7/reference/simple_stmts.html#del). Let's learn the syntax by looking at a simple example:

In [ ]:
d = {'a': 1, 'b': 2, 'c': 3}
del d['a']
print(d)

{'b': 2, 'c': 3}


We can create a function using `del` that will return a copy of our dictionary with the key removed:

In [ ]:
def del_key(dict_, key):
    # create a copy so we don't
    # modify the original dict
    modified_dict = dict_.copy()
    del modified_dict[key]
    return modified_dict

Let's use this function to delete the `createdAtI` key from first_story:

In [ ]:
first_story = del_key(first_story, 'createdAtI')
jprint(first_story)

{
    "author": "dragongraphics",
    "createdAt": "2014-05-29T08:07:50Z",
    "numComments": 0,
    "objectId": "7815238",
    "points": 2,
    "storyText": "",
    "tags": [
        "story",
        "author_dragongraphics",
        "story_7815238"
    ],
    "title": "Are we getting too Sassy? Weighing up micro-optimisation vs. maintainability",
    "url": "http://ashleynolan.co.uk/blog/are-we-getting-too-sassy"
}


The dictionary returned by the function no longer includes the `createdAtI` key.

Let's use a loop and the `del_key()` function to remove the `createdAtI` key from every story in our Hacker News data set

**Instructions:**

We have provided the code for the `del_key()` function.

1. Create an empty list, `hn_clean` to store the cleaned data set.
2. Loop over the dictionaries in the `hn` list. In each iteration:
 - Use the `del_key()` function to delete the `createdAtI` key from the dictionary.
 - Append the cleaned dictionary to `hn_clean`.

In [ ]:
# Answer so the next code works

def del_key(dict_, key):
    # create a copy so we don't
    # modify the original dict
    modified_dict = dict_.copy()
    del modified_dict[key]
    return modified_dict
hn_clean = []

for d in hn:
    new_d = del_key(d, 'createdAtI')
    hn_clean.append(new_d)

## 4. Writing List Comprehensions

Here is the solution code from the previous screen:

In [ ]:
hn_clean = []

for d in hn:
    new_d = del_key(d, 'createdAtI')
    hn_clean.append(new_d)

The task we performed is an extremely common one. Specifically, we:

- Iterated over values in a list.
- Performed a transformation on those values.
- Assigned the result to a new list.

Python includes a special syntax shortcut for tasks that meet these criteria: **List Comprehensions**. A list comprehension provides a concise way of creating lists in a single line of code.



List comprehensions can look complex at first, but we are simply reordering the elements of our for loop code. To keep things simple, we'll start with a basic example, where we want to add 1 to each item in a list of integers.

In [ ]:
ints = [1, 2, 3, 4]

plus_one = []
for i in ints:
    plus_one.append(i + 1)

print(plus_one)

[2, 3, 4, 5]


Let's start by labeling the three main parts of our loop:
![img](https://s3.amazonaws.com/dq-content/355/loop_components.svg)

To transform this structure into a list comprehension, we do the following within brackets:

- Start with the code that transforms each item.
- Continue with our for statement (without a colon).

We can then assign the list comprehension to a variable name. The animation below shows how we convert the manual loop version to a list comprehension.
![img](https://s3.amazonaws.com/dq-content/355/list_comp_anim.svg)

Let's look at a second example, where we want to multiply each item in the list by 10:

In [ ]:
times_ten = []
for i in ints:
    times_ten.append(i * 10)

print(times_ten)

[10, 20, 30, 40]


To convert this to a list comprehension, we follow the same pattern:
![img](https://s3.amazonaws.com/dq-content/355/loop_vs_lc_2.svg)

The "transformation" step of our list comprehension can be anything, including a function or method. In the example below, we are applying a function to a list of floats to round them to integers:

In [ ]:
floats = [2.1, 8.7, 4.2, 8.9]

rounded = []
for f in floats:
    rounded.append(round(f))

print(rounded)

[2, 9, 4, 9]


To convert to a list comprehension, we simply rearrange the components:
![img](https://s3.amazonaws.com/dq-content/355/loop_vs_lc_3.svg)

Just like in a normal loop, we can use any name for our iterator variable. Here, we have used `f`.

For the last example, we'll apply a method to each string in a list to capitalize it. We won't color the different components, so we can get used to how that looks.



In [ ]:
letters = ['a', 'b', 'c', 'd']

caps = []
for l in letters:
    caps.append(l.upper())

Even though we've used a different kind of transformation, the ordering of the list comprehension remains the same:

In [ ]:
caps = [l.upper() for l in letters]
print(caps)

['A', 'B', 'C', 'D']


Let's recap what we have learned so far. A list comprehension can be used where we:

- Iterated over values in a list.
- Performed a transformation on those values.
- Assigned the result to a new list.

To transform a loop to a list comprehension, in brackets we:

- Start with the code that transforms each item.
- Continue with our for statement (without a colon).


We are going to write a list comprehension version of the code from the previous screen. To help, we've provided a copy of the code with the components labeled.

![img](https://s3.amazonaws.com/dq-content/355/loop_components_hn.svg)

**Instructions:**

We've provided the solution for the previous screen in comments for you to use as a reference.

1. Create a list comprehension representation of the loop from the previous screen:
 - Call the `del_key()` function to remove the `createdAtI` value from each dictionary in the `hn` list.
 - Assign the results to a new list, `hn_clean`.

## 5. Using List Comprehensions to Transform and Create Lists

List comprehensions can be used for many different things. Three common applications are:

1. Transforming a list
2. Creating a new list
3. Reducing a list

On this screen, we're going to look at the first two of these applications.

The first application, **transforming a list**, is the category that all the examples you've seen so far fit under. You are taking an existing list, applying a transformation to every value, and assigning it to a variable.



The example below uses a list comprehension to transform a list of square numbers into their "square roots":

![img](https://s3.amazonaws.com/dq-content/355/lc_application_1.svg)

The second application, **creating a new list**, is useful for creating test data or data that is based on a set of numbers.

As an example, let's create a list of generic columns names that we could use to create a dataframe using the `range()` [function](https://docs.python.org/3/library/stdtypes.html#range) and the `str.format()` [method](https://docs.python.org/3/library/stdtypes.html#str.format) to combine numbers and text:

![img](https://s3.amazonaws.com/dq-content/355/lc_application_2_2.svg)


We can then use this to create an empty dataframe with labels:

```
cols = ["col_{}".format(i) for i in range(1,5)]
data = np.zeros((4,4))

df = pd.DataFrame(data, columns=cols)
print(df)
```



```
_  col_1  col_2  col_3  col_4
0    0.0    0.0    0.0    0.0
1    0.0    0.0    0.0    0.0
2    0.0    0.0    0.0    0.0
3    0.0    0.0    0.0    0.0
```

Let's use list comprehension to create a new list containing just the URLs from each story.

**Instructions:**

Use a list comprehension to extract the url value from each dictionary in hn_clean. Assign the result to urls.

In [ ]:
urls = [d['url'] for d in hn_clean]

## 6. Using List Comprehensions to Reduce a List

The last common application of list comprehensions is **reducing a list**. Let's say we had a list of integers and we wanted to remove any integers that were smaller than 50. We could do this by adding an if statement to our loop:
![img](https://s3.amazonaws.com/dq-content/355/lc_application_3_1.svg)

Our loop has one new component — the if statement, which we've colored yellow. Notice that instead of a transformation, we have just the list item itself (`i`) in red. Both if statements and transformations are optional in list comprehensions, but we must include some value to populate the elements in the new list.

To include an if statement in a list comprehension, we include it at the very end, before the closing bracket:

![img](https://s3.amazonaws.com/dq-content/355/lc_application_3_2.svg)


We can use this technique to quickly and easily filter our data set using an if statement. Let's look at how we could use this to quickly count the number of stories that have comments. We'll start with a version using a loop:

In [ ]:
has_comments = []

for d in hn_clean:
    if d['numComments'] > 0:
        has_comments.append(d)

num_comments = len(has_comments)
print(num_comments)

9279


Now, let's use a list comprehension to perform the same calculation:

In [ ]:
has_comments = [d for d in hn_clean if d['numComments'] > 0]

num_comments = len(has_comments)
print(num_comments)

9279


Let's use list comprehension to count how many stories have more than 1,000 points.

**Instructions:**

1. Use list comprehension to create a new list, `thousand_points`:
 - The list should contain values from `hn_clean` where the `points` key has a value greater than 1000.
2. Count the number of values in `thousand_points` and assign the result to `num_thousand_points`.

## 7. Passing Functions as Arguments

In previous missions, we learned to use Python's built-in functions to analyze data in lists, like `min()`, `max()`, and `sorted()`.

What if we wanted to use these functions to work with data in JSON form? Let's use our demo JSON object to try and see what happens. First, we'll quickly remind ourselves of the data:

In [ ]:
jprint(json_obj)

[
    {
        "age": 36,
        "favorite_foods": [
            "Pumpkin",
            "Oatmeal"
        ],
        "name": "Sabine"
    },
    {
        "age": 40,
        "favorite_foods": [
            "Chicken",
            "Pizza",
            "Chocolate"
        ],
        "name": "Zoe"
    },
    {
        "age": 40,
        "favorite_foods": [
            "Caesar Salad"
        ],
        "name": "Heidi"
    }
]


Let's try and use Python to return the dictionary of the person with the lowest age:

In [ ]:
min(json_obj)

TypeError: ignored

We get an error because Python doesn't have any way to tell whether one dictionary object is "greater" than another.

There is a way we can actually tell functions like `min()`, `max()`, and `sorted()` how to sort complex objects like dictionaries and lists of lists. We do this by using the optional `key` argument. The official Python documentation contains the following excerpts that describe how the argument works:


> "The `key` argument specifies a one-argument ordering function like that used for `list.sort()`. `key` specifies a function of one argument that is used to extract a comparison key from each list element. The key corresponding to each item in the list is calculated once and then used for the entire sorting process."

These excerpts tell us we need to specify a function as an argument to control the comparison between values. Up until now, we've only passed variables or values as arguments, but not functions!

We'll learn the specifics of this particular application in a moment, but for now, we're going to explore how to pass a function as an argument.

Let's define a very simple function as an example:

In [ ]:
def greet():
    return "hello"

greet()

If we try to examine the type of our function, we are unsuccessful:

In [ ]:
t = type(greet())
print(t)

What happens is that `greet()` is executed first; it returns the string `"hello"`, and then the `type()` function tells us the type of that string:

![img](https://s3.amazonaws.com/dq-content/355/type_of_func_1.svg)

We need to find a way to look at the function itself, rather than the result of the function. The key to this is the parentheses: `()`.

The parentheses are what tells Python to execute the function, so if we **omit the parentheses we can treat a function like a variable**, rather than working with the output of the function:

In [ ]:
t = type(greet)
print(t)

There are other variable-like behaviors we can also use when we omit the parentheses from a function. For instance, we can assign a function to a new variable name:



In [ ]:
greet_2 = greet

greet_2()

Now that we understand how to treat a function as variable, let's look at how we can run a function inside another function by passing it as an argument:

In [ ]:
def run_func(func):
    print("RUNNING FUNCTION: {}".format(func))
    return func()

run_func(greet)

![img](https://s3.amazonaws.com/dq-content/355/func_as_arg.svg)
Now that we have some intuition on how to pass functions as arguments, let's see how we use a function to control the behavior of the `sorted()` function:
![img](https://s3.amazonaws.com/dq-content/355/sorted_key.svg)

Let's look at the same thing in code form:


In [ ]:
def get_age(json_dict):
    return json_dict['age']

youngest = min(json_obj, key=get_age)
jprint(youngest)

Let's use this technique to find the story that has the greatest number of comments.

**Instructions:**

1. Create a "key function" that accepts a single dictionary and returns the value from the `numComments` key.
2. Use the `max()` function with the "key function" you just created to find the value from the `hn_clean` list with the most comments:
 - Assign the result to the variable `most_comments`.

## 8. Lambda Functions

Usually, we create functions when we want to perform the same task many times. In the previous exercise, we created a function to use just once — as an argument to `max()`.

Python provides a special syntax to create temporary functions for situations like these. These functions are called **lambda functions**. Lambda functions can be defined in a single line, which allows you to define a function you want to pass as an argument at the time you need it.

While it's unusual to assign a lamdba function to a variable name, we'll do that while we learn lambda functions through some simple examples. We'll start with a function that returns a single argument without modifying it:



In [ ]:
def unchanged(x):
    return x

Let's give each component of the function a name so we can more easily talk about it:
![img](https://s3.amazonaws.com/dq-content/355/lambda_1_components.svg)

We're calling the returned element "transformation," even though there is no transformation. This will make sense as we introduce more complex examples.

To create a lambda function equivalent of this function, we:

- Use the lambda keyword, followed by
- The parameter and a colon, and then
- The transformation we wish to perform on our argument

We can then assign that to the function name:
![img](https://s3.amazonaws.com/dq-content/355/lambda_1_anim.svg)


Let's look at a comparison of the original function and its lamdba equivalent:
![img](https://s3.amazonaws.com/dq-content/355/lambda_1_comparison.svg)

Now let's look at a second example, where we add a simple transformation to the argument:

![img](https://s3.amazonaws.com/dq-content/355/lambda_2_components.svg)



We follow exactly the same steps to make a lambda function:

![img](https://s3.amazonaws.com/dq-content/355/lambda_2_comparison.svg)

If we want to create a lambda that has multiple parameters, we follow exactly the same steps:

![img](https://s3.amazonaws.com/dq-content/355/lambda_3_components.svg)

## 9. Using Lambda Functions to Analyze JSON data

## 10. Reading JSON files into pandas

## 11. Exploring Tags Using the Apply Function

## 12. Extracting Tags Using Apply with a Lambda Function